scapy/arch/linux/__init__.py

```diff
@@ -332,39 +352,68 @@ def recv(self, x=MTU, **kwargs):
 
     def send(self, x):
         # type: (Packet) -> int
+        # Select the file descriptor to send the packet on.
         iff = x.route()[0]
         if iff is None:
             iff = network_name(conf.iface)
-        sdto = (iff, self.type)
-        self.outs.bind(sdto)
-        sn = self.outs.getsockname()
-        ll = lambda x: x  # type: Callable[[Packet], Packet]
         type_x = type(x)
-        if type_x in conf.l3types:
-            sdto = (iff, conf.l3types.layer2num[type_x])
-        if sn[3] in conf.l2types:
-            ll = lambda x: conf.l2types.num2layer[sn[3]]() / x
-        if self.lvl == 3 and not issubclass(self.LL, type_x):
-            warning("Incompatible L3 types detected using %s instead of %s !",
-                    type_x, self.LL)
-            self.LL = type_x
-        sx = raw(ll(x))
-        x.sent_time = time.time()
+        if iff not in self.send_socks:
+            self.send_socks[iff] = L3PacketSocket(
+                iface=iff,
+                type=conf.l3types.layer2num.get(type_x, self.type),
+                filter=self.filter,
+                promisc=self.promisc,
+            )
+        fd = self.send_socks[iff].outs
+        if self.lvl == 3:
+            if not issubclass(self.LL, type_x):
+                warning("Incompatible L3 types detected using %s instead of %s !",
+                        type_x, self.LL)
+                self.LL = type_x
+        if self.lvl == 2:
+            sx = bytes(self.LL() / x)
+        else:
+            sx = bytes(x)
+        # Now send.
+        try:
+            x.sent_time = time.time()
+        except AttributeError:
+            pass
         try:
-            return self.outs.sendto(sx, sdto)
+            return fd.send(sx)
         except socket.error as msg:
             if msg.errno == 22 and len(sx) < conf.min_pkt_size:
-                return self.outs.send(
+                return fd.send(
                     sx + b"\x00" * (conf.min_pkt_size - len(sx))
                 )
             elif conf.auto_fragment and msg.errno == 90:
                 i = 0
                 for p in x.fragment():
-                    i += self.outs.sendto(raw(ll(p)), sdto)
+                    i += fd.send(bytes(self.LL() / p))
                 return i
             else:
                 raise
 
+    @staticmethod
+    def select(sockets, remain=None):
```

Comment:
Could we use a generic `select()`method instead of duplicating it in all sockets?

Reply:
I'm not sure if that's clearer, considering the types change in each function.
We would require a dependency to another `L3Socket` or something. I think sockets are already abstracted enough

Reply:
A `classmethod` could do the trick. What do you think?

Reply:
sounds like a LOT of refactoring work that I honnestly don't really want to do :(